libtensorflow_cc.so: Undefined references due to static dependencies with cc_shared_library and symbol hiding by linker scripts

### Issue Type

Build/Install

### Source

source

### Tensorflow Version

617ca3da6c56ff668c202f853757d8df6ac779d7

### Custom Code

No

### OS Platform and Distribution

Linux Ubuntu 20.04

### Mobile device

_No response_

### Python version

3.8.10

### Bazel version

5.1.1 and 6.0.0-pre.20220421.3

### GCC/Compiler version

gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0

### CUDA/cuDNN version

11.7

### GPU model and memory

3080

### Current Behaviour?

Since the recent change to use `cc_shared_library`, I am seeing many undefined reference linker errors coming from LLVM/MLIR and libjpeg.

Both `libtensorflow_framework` and `libtensorflow_cc` have LLVM and libjpeb as static dependencies. Since `libtensorflow_cc` has a dynamic dependency on `libtensorflow_framework`, it appears that the behavior of `cc_shared_library` is to retrieve the static deps via `libtensorflow_framework` (previously, both libraries would duplicate them). The problem is that it cannot load the symbols because the [linker](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tf_private_symbols.lds) [scripts](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tf_framework_version_script.lds) are explicitly hiding those exact symbols.

### Standalone code to reproduce the issue

Simply build `libtensorflow_cc.so`.

```shell
docker pull tensorflow/tensorflow:devel
docker run -it -w /tensorflow_src -v $PWD:/mnt -e HOST_PERMS="$(id -u):$(id -g)" tensorflow/tensorflow:devel bash
git pull
echo "" | ./configure
bazel build --config=opt  --experimental_ui_max_stdouterr_bytes=-1 //tensorflow:libtensorflow_cc.so
```

` --experimental_ui_max_stdouterr_bytes=-1` is only needed to see the error message, otherwise it is hidden because it is too long due to all of the undefined references.

### Relevant log output


There are thousands of undefined references to LLVM,MLIR and libjpeg, here is a small snippet.

```shell
ERROR: /opt/tensorflow/tensorflow-source/tensorflow/BUILD:1135:20: Linking tensorflow/libtensorflow_cc.so.2.10.0 failed: (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command (from target //tensorflow:libtensorflow_cc.so.2.10.0) external/local_co[294/1887]crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc @bazel-out/k8-opt/bin/tensorflow/libtensorflow_cc.so.2.10.0-2.params                                                                                                                                                            /usr/bin/ld: bazel-out/k8-opt/bin/tensorflow/core/lib/jpeg/libjpeg_internal.pic.a(jpeg_mem.pic.o): in function `tensorflow::jpeg::GetImageInfo(void const*, int, int*, int*, int*)':                                                                                                    jpeg_mem.cc:(.text._ZN10tensorflow4jpeg12GetImageInfoEPKviPiS3_S3_+0x95): undefined reference to `jpeg_std_error'                                                                                                                                                                       /usr/bin/ld: jpeg_mem.cc:(.text._ZN10tensorflow4jpeg12GetImageInfoEPKviPiS3_S3_+0xe2): undefined reference to `jpeg_CreateDecompress'                                                                                                                                                   /usr/bin/ld: jpeg_mem.cc:(.text._ZN10tensorflow4jpeg12GetImageInfoEPKviPiS3_S3_+0x107): undefined reference to `jpeg_read_header'                                                                                                                                                       /usr/bin/ld: jpeg_mem.cc:(.text._ZN10tensorflow4jpeg12GetImageInfoEPKviPiS3_S3_+0x10f): undefined reference to `jpeg_calc_output_dimensions'                                                                                                                                            /usr/bin/ld: jpeg_mem.cc:(.text._ZN10tensorflow4jpeg12GetImageInfoEPKviPiS3_S3_+0x157): undefined reference to `jpeg_destroy_decompress'                                                                                                                                                /usr/bin/ld: bazel-out/k8-opt/bin/tensorflow/core/lib/jpeg/libjpeg_internal.pic.a(jpeg_mem.pic.o): in function `tensorflow::jpeg::(anonymous namespace)::UncompressLow(void const*, tensorflow::jpeg::(anonymous namespace)::FewerArgsForCompiler*)':  
```

--- a/.bazelrc
+++ b/.bazelrc
@@ -150,7 +150,7 @@ build --define=no_hdfs_support=true
 build --experimental_cc_shared_library
 
 # cc_shared_library ensures no library is linked statically more than once.
-build --experimental_link_static_libraries_once=true
+build --experimental_link_static_libraries_once=false
 
 # Default options should come above this line.
 

--- a/tensorflow/BUILD
+++ b/tensorflow/BUILD
@@ -995,7 +995,6 @@ tf_cc_shared_library(
         "//tensorflow:tf_framework_version_script.lds",
         "//tensorflow:tf_private_symbols.lds",
     ],
-    exports_filter = PACKAGE_STATIC_DEPS,
     framework_so = [],
     linkopts = select({
         "//tensorflow:macos": [

